File: lib/to_mdx/parsers/open_document/style/marker.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Style
      class Marker < Base
        attr_reader :marker_name, :view_box, :path_data

        def initialize(element = nil, context = {})
          super(element, context)
          @marker_name = element&.attributes&.[]("draw:name")
          @view_box = element&.attributes&.[]("svg:viewBox")
          @path_data = element&.attributes&.[]("svg:d")
        end

        def to_css
          return {} unless @marker_name

          properties = extract_properties
          return {} if properties.empty?

          css_properties = generate_marker_properties(properties)
          css_properties.any? ? { @marker_name => css_properties } : {}
        end

        def extract_properties
          return {} unless @element&.attributes

          @element.attributes.each_with_object({}) do |(name, value), props|
            props[name] = value
          end
        end

        private

        def generate_marker_properties(properties)
          css_properties = {}

          properties.each do |name, value|
            case name
            when "svg:viewBox"
              css_properties["viewBox"] = value
            when "svg:d"
              css_properties["d"] = value
            end
          end

          css_properties
        end

        register_for "draw:marker"
      end
    end
  end
end

File: src/parsers/open-document/style/marker.ts
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class Marker extends BaseElement {
  private markerName: string | null = null;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.markerName = this.attr("draw:name");
  }

  toMdx(): string {
    return "";
  }

  toString(): string {
    return "";
  }

  toCss(): Record<string, Record<string, string>> {
    if (!this.markerName) {
      return {};
    }

    const properties = this.extractProperties();
    if (Object.keys(properties).length === 0) {
      return {};
    }

    const cssProperties = this.generateMarkerProperties(properties);
    return Object.keys(cssProperties).length > 0 ? { [this.markerName]: cssProperties } : {};
  }

  private extractProperties(): Record<string, string> {
    const properties: Record<string, string> = {};

    Array.from(this.element.attributes).forEach((attr) => {
      properties[attr.name] = attr.value;
    });

    return properties;
  }

  private generateMarkerProperties(properties: Record<string, string>): Record<string, string> {
    const cssProperties: Record<string, string> = {};

    Object.entries(properties).forEach(([name, value]) => {
      switch (name) {
        case "svg:viewBox":
          cssProperties.viewBox = value;
          break;
        case "svg:d":
          cssProperties.d = value;
          break;
      }
    });

    return cssProperties;
  }

  static {
    BaseElement.registerFor.call(this, "draw:marker");
  }
}
